documentation/writing_tests.md

```diff
@@ -0,0 +1,949 @@
+      env.AZURE_TENANT_ID,
+      env.AZURE_CLIENT_ID,
+      env.AZURE_CLIENT_SECRET
+    );
+
+    const keyVaultUrl = "https://myKeyVault.vault.azure.net";
+    client = new KeysClient(keyVaultUrl, credential);
+  });
+
+  afterEach(function () {
+    recorder.stop();
+  });
+```
+
+### Preparing some of the test cases
+
+Some test cases might need some common setup to properly work. If at least more than one of these tests exist in a single file, their setup can be moved out of each test case into a function (or outside constant), as long as the function (or constant) is declared inside of that same test file, and as long as the common code is sufficiently large. For readability purposes, if the common code is less than four lines long (as an arbitrary reasonable estimate), it will be clearer if it's left repeated in each test case.
+
+Let's define an example in which a SDK client might need several lines of setup in two specific tests:
+
+```ts
+describe("some group of functionalities", function() {
+  let client;
+
+  beforeEach(function() {
+    client = new Client();
+  });
+
+  it("should test A with a proper setup", function() {
+    client.prepare({
+      propertyA: 1,
+      propertyB: 2,
+      propertyC: 3,
+      propertyD: 4,
+    });
+
+    const result = await client.A();
+    assert.ok(result.value);
+  });
+
+  it("should test B with a proper setup", function() {
+    client.prepare({
+      propertyA: 1,
+      propertyB: 2,
+      propertyC: 3,
+      propertyD: 4,
+    });
+
+    const result = await client.B();
+    assert.ok(result.value);
+  });
+});
+```
+
+It is possible to move the common setup code to a function in the same file, as in the following example code:
+
+```ts
+describe("some group of functionalities", function() {
+  let client: Client;
+
+  beforeEach(function() {
+    client = new Client();
+  });
+
+  function defaultPrepareClient(client: Client) {
+    client.prepare({
+      propertyA: 1,
+      propertyB: 2,
+      propertyC: 3,
+      propertyD: 4,
+    });
+  }
+
+  it("should test A with a proper setup", function() {
+    defaultPrepareClient(client);
+    const result = await client.A();
+    assert.ok(result.value);
+  });
+
+  it("should test B with a proper setup", function() {
+    defaultPrepareClient(client);
+    const result = await client.B();
+    assert.ok(result.value);
+  });
+});
+```
+
+### Universal utilities
+
+In case that a specific utility function might be useful for more than one test case, and more than one test file, and even potentially in more than one project, this code should be moved away from each client into a common private package in the [test-utils](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/) folder. If the package doesn't exist, it should be created.
+
+Let's say we have a test utility that we call `retry.ts`, with the following content:
+
+```ts
+import { delay as coreDelay } from "@azure/core-http";
```

Comment:
Note that in another librarywe're using 'delay' from test-utils. Is there any reason to use one vs the other when it comes to writing test code?

Reply:
I will make this use the recorder's delay. Thanks!

Reply:
Done!